src/VisualStudio/Core/Def/Implementation/ProjectSystem/RuleSets/VisualStudioRuleSetManager.RuleSetFile.cs

```diff
@@ -11,60 +11,72 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
 {
     internal sealed partial class VisualStudioRuleSetManager
     {
-        private sealed class RuleSetFile : IRuleSetFile
+        private sealed class RuleSetFile : IRuleSetFile, IDisposable
         {
-            private readonly string _filePath;
-            private readonly List<FileChangeTracker> _trackers;
             private readonly VisualStudioRuleSetManager _ruleSetManager;
+            private readonly object _gate = new object();
+
+            /// <summary>
+            /// The list of file trackers we have. This is null if we haven't even computed trackers yet; but can be empty if we
+            /// already know the file was modified and we unsubscribed.
+            /// </summary>
+            private List<FileChangeTracker> _trackers;
```

Comment:
:bulb: Consider using `ImmutableArray<FileChangeTracker>` or `ImmutableList<FileChangeTracker>` with atomic exchange. The former offers better efficiency if the list is typically constructed from empty, while the latter offers efficiency if the list is typically incrementally populated.

Reply:
We directly create the type with the known capacity, so that'll be fastest of all, right?

Reply:
Ah, this wasn't a :bulb: for efficiency. The immutable collections allow some flexibility with respect to hoisting cleanup operations outside the locks (you can exchange the list with an empty list, then use the local copy to clean things up).

Reply:
Aiming for KISS here.